src/github.com/couchbase/sync_gateway/base/sharded_sequence_clock.go

```diff
@@ -0,0 +1,290 @@
+func (scp *ShardedClockPartition) Marshal() ([]byte, error) {
+	var output bytes.Buffer
+	enc := gob.NewEncoder(&output)
+	err := enc.Encode(scp)
+	if err != nil {
+		return nil, err
+	}
+	return output.Bytes(), nil
+}
+
+func (scp *ShardedClockPartition) Unmarshal(value []byte) error {
+	input := bytes.NewBuffer(value)
+	dec := gob.NewDecoder(input)
+	err := dec.Decode(&scp)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (scp *ShardedClockPartition) SetSequence(vb uint16, seq uint64) {
+	scp.Values[vb] = seq
+	scp.dirty = true
+}
+
+func NewShardedClock(key string, partitions *IndexPartitions, bucket Bucket) *ShardedClock {
+	clock := &ShardedClock{
+		baseKey:      key,
+		partitionMap: partitions,
+		countKey:     fmt.Sprintf(kCountKeyFormat, key),
+		bucket:       bucket,
+	}
+
+	// Initialize partitions
+	numPartitions := len(partitions.PartitionDefs)
+
+	clock.partitions = make(map[uint16]*ShardedClockPartition, numPartitions)
+	clock.partitionKeys = make([]string, numPartitions)
+
+	return clock
+
+}
+
+func NewShardedClockWithPartitions(key string, partitions *IndexPartitions, bucket Bucket) *ShardedClock {
+	clock := &ShardedClock{
+		baseKey:      key,
+		partitionMap: partitions,
+		countKey:     fmt.Sprintf(kCountKeyFormat, key),
+		bucket:       bucket,
+	}
+
+	// Initialize partitions
+	numPartitions := len(partitions.PartitionDefs)
+
+	clock.partitions = make(map[uint16]*ShardedClockPartition, numPartitions)
+	clock.partitionKeys = make([]string, numPartitions)
+
+	// Initialize empty clock partitions
+	for _, partitionDef := range partitions.PartitionDefs {
+		clock.partitions[partitionDef.Index] = NewShardedClockPartition(key, partitionDef.Index)
+		clock.partitionKeys[partitionDef.Index] = clock.partitions[partitionDef.Index].Key
+	}
+
+	return clock
+
+}
+
+func (s *ShardedClock) Write() (err error) {
+	// write all modified partitions to bucket
+	shardedClockExpvars.Add("count_write", 1)
+	var wg sync.WaitGroup
+	for _, partition := range s.partitions {
+		if partition.dirty {
+			wg.Add(1)
+			// TODO: replace with SetBulk when available
+			go func(p *ShardedClockPartition) {
+				defer wg.Done()
+				value, err := p.Marshal()
+				casOut, err := s.bucket.WriteCas(p.Key, 0, 0, p.cas, value, sgbucket.Raw)
+
+				if err != nil {
+					Warn("Error writing sharded clock partition [%d]:%v", p.Key, err)
+					shardedClockExpvars.Add("partition_cas_failures", 1)
+					return
+				}
+
+				p.cas = casOut
+				p.dirty = false
+			}(partition)
+		}
+	}
+	wg.Wait()
+
+	// Increment the clock count
+	s.counter, err = s.bucket.Incr(s.countKey, 1, 0, 0)
+	return err
+}
+
+// Loads clock from bucket.  If counter isn't changed, returns false and leaves as-is.
+// For newly initialized ShardedClocks (counter=0), this will only happen if there are no
+// entries in the bucket for the clock.
+func (s *ShardedClock) Load() (isChanged bool, err error) {
+
+	newCounter, err := s.bucket.Incr(s.countKey, 0, 0, 0)
```

Comment:
Incr works as-is when using walrus/go-couchbase.  Doing a Get against walrus will throw an error (as it doesn't do automatic transcoding to uint).